lib/NotificationGenerator.php

```diff
@@ -0,0 +1,130 @@
+ *
+ */
+
+namespace OCA\Activity;
+
+use OCP\Activity\IEvent;
+use OCP\Activity\IManager as ActivityManager;
+use OCP\Notification\IManager as NotificationManager;
+use OCP\IL10N;
+use OCP\Notification\INotification;
+use OCP\Notification\INotifier;
+
+class NotificationGenerator implements INotifier {
+	protected $data;
+	protected $activityManager;
+	protected $userSettings;
+	protected $l10n;
+	protected $notificationManager;
+
+	public function __construct(Data $data, ActivityManager $activityManager, NotificationManager $notificationManager, UserSettings $userSettings, IL10N $l10n) {
+		$this->data = $data;
+		$this->activityManager = $activityManager;
+		$this->notificationManager = $notificationManager;
+		$this->userSettings = $userSettings;
+		$this->l10n = $l10n;
+	}
+
+	public function sendNotificationForEvent(IEvent $event, int $activityId) {
+		$selfAction = $event->getAffectedUser() === $event->getAuthor();
+		$notifySetting = $this->userSettings->getUserSetting($event->getAffectedUser(), 'notify', $event->getType());
+
+		// User is not the author or wants to see their own actions
+		$createStream = !$selfAction || $this->userSettings->getUserSetting($event->getAffectedUser(), 'setting', 'self');
+
+		if (($notifySetting || true) && $event->getGenerateNotification()) {
+			$this->notificationManager->notify($this->getNotificationForEvent($event, $activityId));
+		}
+	}
+
+	private function getNotificationForEvent(IEvent $event, int $activityId): INotification {
+
+		$notification = $this->notificationManager->createNotification();
+		$notification->setApp($event->getApp());
+		$notification->setUser($event->getAffectedUser());
+		$notification->setDateTime(\DateTime::createFromFormat('U', (string)$event->getTimestamp()));
+		$notification->setObject('activity_notification', (string)$activityId);
+		$notification->setSubject($event->getSubject());
+
+		if ($notification->getRichMessage()) {
+			$notification->setRichMessage($event->getRichMessage(), $event->getRichMessageParameters());
+		}
+
+		if ($event->getMessage()) {
+			$notification->setMessage($event->getMessage());
+		}
+
+		return $notification;
+	}
+
+	private function populateEvent(IEvent $event, string $language) {
+		foreach ($this->activityManager->getProviders() as $provider) {
+			try {
+				$this->activityManager->setFormattingObject($event->getObjectType(), $event->getObjectId());
+				$event = $provider->parse($language, $event);
+			} catch (\InvalidArgumentException $e) {
+			}
+		}
+
+		return $event;
+	}
+
+	public function getID(): string {
+		return 'activity';
+	}
+
+	public function getName(): string {
+		return 'Activity';
+	}
+
+	public function prepare(INotification $notification, string $languageCode): INotification {
+		if ($notification->getObjectType() !== 'activity_notification') {
+			throw new \InvalidArgumentException();
+		}
+
+		$event = $this->data->getById((int)$notification->getObjectId());
```

Comment:
Shouldn't happen, but who knows, so:
Bonus points for checking notification user === activity affected user